staging/src/k8s.io/apiserver/pkg/server/config.go

```diff
@@ -128,6 +129,8 @@ type Config struct {
 
 	// BuildHandlerChainFunc allows you to build custom handler chains by decorating the apiHandler.
 	BuildHandlerChainFunc func(apiHandler http.Handler, c *Config) (secure http.Handler)
+	// HandlerChainWaitGroup allows you to wait for all chain handlers exit after the server shutdown.
```

Comment:
do we really need this in the config? I would just create it in the GenericApiServer object.

Reply:
ok, originally, I just thought it as BuildHandlerChainFunc AuditBackend, etc.

And i want to know what's the difference between GenericAPIServer and Config, they have many same fields. What's the design intention? I am confused.
```
	s := &GenericAPIServer{
		discoveryAddresses:     c.DiscoveryAddresses,
		LoopbackClientConfig:   c.LoopbackClientConfig,
		legacyAPIGroupPrefixes: c.LegacyAPIGroupPrefixes,
		admissionControl:       c.AdmissionControl,
		requestContextMapper:   c.RequestContextMapper,
		Serializer:             c.Serializer,
		AuditBackend:           c.AuditBackend,
		delegationTarget:       delegationTarget,
		HandlerChainWaitGroup:  c.HandlerChainWaitGroup,

		minRequestTimeout: time.Duration(c.MinRequestTimeout) * time.Second,

		SecureServingInfo: c.SecureServingInfo,
		ExternalAddress:   c.ExternalAddress,

		Handler: apiServerHandler,

		listedPathProvider: apiServerHandler,

		swaggerConfig: c.SwaggerConfig,
		openAPIConfig: c.OpenAPIConfig,

		postStartHooks:         map[string]postStartHookEntry{},
		preShutdownHooks:       map[string]preShutdownHookEntry{},
		disabledPostStartHooks: c.DisabledPostStartHooks,

		healthzChecks: c.HealthzChecks,

		DiscoveryGroupManager: discovery.NewRootAPIsHandler(c.DiscoveryAddresses, c.Serializer, c.RequestContextMapper),

		enableAPIResponseCompression: c.EnableAPIResponseCompression,
	}
```

Reply:
But if I do this, may introduce `func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) ` modification, because here I must acquire waitGroup from genericApiserver. So, is it worth the complexity?

Reply:
Yes, we need. Because if only reserve it in GenericApiServer . It will introduce the change here
```
handlerChainBuilder := func(handler http.Handler) http.Handler {
		return c.BuildHandlerChainFunc(handler, c.Config)
	}
	apiServerHandler := NewAPIServerHandler(name, c.RequestContextMapper, c.Serializer, handlerChainBuilder, delegationTarget.UnprotectedHandler())

	s := &GenericAPIServer{
```

Reply:
HandlerChainWaitGroup now both used in Config and GenericAPIServer.